astroquery/mast/observations.py

```diff
@@ -0,0 +1,902 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+MAST Observations
+=================
+
+This module contains various methods for querying MAST observations.
+"""
+
+
+import warnings
+import json
+import time
+import os
+import uuid
+
+import numpy as np
+
+from requests import HTTPError
+
+import astropy.units as u
+import astropy.coordinates as coord
+
+from astropy.table import Table, Row, vstack, MaskedColumn
+from astropy.logger import log
+
+from astropy.utils import deprecated
+from astropy.utils.console import ProgressBarOrSpinner
+from astropy.utils.exceptions import AstropyDeprecationWarning
+
+from six.moves.urllib.parse import quote as urlencode
+
+from ..query import QueryWithLogin
+from ..utils import commons, async_to_sync
+from ..utils.class_or_instance import class_or_instance
+from ..exceptions import (TimeoutError, InvalidQueryError, RemoteServiceError,
+                          ResolverError, MaxResultsWarning,
+                          NoResultsWarning, InputWarning, AuthenticationWarning)
+
+from . import conf, utils, fpl
+from .auth import MastAuth
+from .cloud import CloudAccess
+from .discovery_portal import PortalAPI
+
+
+__all__ = ['Observations', 'ObservationsClass']
+
+
+@async_to_sync
+class ObservationsClass(QueryWithLogin):
+    """
+    MAST Observations query class.
+
+    Class for querying MAST observational data.
+    """
+
+    def __init__(self, mast_token=None):
```

Comment:
I'm wondering whether `__init__()`, `_login()`, `get_token()`, `session_info()`, `logout()` or `_parse_result()` are any different than in `MastClass`?  If not (and if I'm reading it right), then that seems like too much code to duplicate.  I can see a couple options and am open to others.

1)  Extract a common base class that includes those methods (or whatever the actual common behavior is).  Maybe call it something like `MastQueryWithLogin`?

2) Go back to having `ObservationClass` extend `MastClass`.  I know this defeats some of the purpose of this refactoring, but since `MastClass.resolve_object()` is just a pass through to `utils` (and maybe not needed?), `MastClass.service_request_async()` is the only thing `MastClass` has that `ObservationClass` doesn't.

Reply:
To a certain extent I think we have to live with more code reuse than we had before in order to make the code more extendable. And some of it also comes from the goal of leaving the interface untouched from a user perspective.

The `utils.resolve_object` for instance.  Ultimately I think what we want to do is deprecate it from all the classes and use only the utils version, but we can't just get rid of it. Last time I messed with that function it broke lightkurve, so we'll need to have an actual deprecation period and probably open an issue over at lightkurve (and maybe eleanor) to make sure that doesn't happen again.

In terms of the massive overlap between `MastClass` and `ObservationsClass`. I don't want to make `ObservationsClass` a subclass of `MastClass` because the point of separating out the portal connection was so that `ObservationsClass` can also have microservice connections in future. Something that I did debate though was whether to separate `MastClass` from the `PortalAPI` class at all. Now might be a time to revisit that idea.  I could move the `MastClass` functionality into `PortalAPI` and then just define `MastClass = PortalAPI` and `Mast = PortalApi()`. It wouldn't completely eliminate the redundancy, but it would lower it. 

Another driver of the code reuse is that I've based all of the classes on the Astroquery base classes. I did this because I wanted to use the functionality you get from that, and be consistent with the Astroquery design overall, but it does also mean we need to have some of the same functions in all the classes. I'd be interested in any opinions on this as a design choice.

What do you and @phicc  think?

Reply:
I agree that it makes sense for `ObservationClass` not to inherit from `MastClass` for the reason you state.  It does seems like too much duplicate code to me though, so my other thought was to extract the commonality into a new class from which they both inherit.  

That approach makes the most sense if that commonality is expected to persist once `ObservationClass` starts getting the extensions you mention and if the commonality neatly fits a single concept or functionality.  In this case I was wondering if that commonality was the login capability, in which case it might make sense to add a class that inherits from `QueryWithLogin` and adds MAST login capabilities.  It could then be extended by both `ObservationClass` and `MastClass`.